### PR TITLE
feat: [PIPE-22045]: file explorer in gitness

### PR DESCRIPTION
**Pull request by shaurya-harness**
Adds Link to BreadCrumbs shown for Repo Code file destinations
Adds File Explorer sort and display folders first and then files
Adds File Content Viewer that uses monaco code editor to display content and lang-map to provide the proper language as per the filetype

Code View Content
Screenshot 2024-10-10 at 11 07 44 PM

Markdown Preview
Screenshot 2024-10-10 at 11 08 23 PM

### CODE REVIEW

File: apps/gitness/src/utils/git-utils.ts [hunk 1]
```diff
@@ -1,5 +1,131 @@
...
+ * @param integer precision Number of digits after the decimal separator
+ * @return string
+ * @link https://stackoverflow.com/a/18650828/1114931
+ */
+export function formatBytes(bytes: number, decimals = 2) {
+  if (!+bytes) return '0 Bytes'
+
+  const k = 1024
+  const dm = decimals < 0 ? 0 : decimals
+  const sizes = ['Bytes', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB']
```
Comment (praneshg239) [inline apps/gitness/src/utils/git-utils.ts:124]: NIT: do we need to show as KiB or KB format? Which one is more universal?

**Review by praneshg239 — APPROVED** [review]

File: apps/gitness/src/utils/git-utils.ts (hunk 1 above)
Comment (vardanbansal-harness) [reply apps/gitness/src/utils/git-utils.ts:124]: Agree, probably just use `KB` format.

File: apps/gitness/src/components/FileViewer.tsx [hunk 2]
```diff
@@ -13,12 +24,37 @@ import { useGetRepoRef } from '../framework/hooks/useGetRepoPath'
 import { getTrimmedSha, normalizeGitRef } from '../utils/git-utils'
 import { PathParams } from '../RouteDefinitions'
 import { timeAgoFromISOTime } from '../pages/pipeline-edit/utils/time-utils'
+import FileContentViewer from './FileContentViewer'
+
+interface PathParts {
+  path: string
+  parentPath: string
+}
+function splitPathWithParents(fullResourcePath: string) {
```
Comment (cjlee01) [inline apps/gitness/src/components/FileViewer.tsx:33]: can we move this funct to a utils function?

File: apps/gitness/src/components/FileContentViewer.tsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  const themeConfig = useMemo(
+    () => ({
+      defaultTheme: 'dark',
+      themes
+    }),
+    []
+  )
+
+  useEffect(() => {
+    setView(setDefaultView(language))
```
Comment (cjlee01) [inline apps/gitness/src/components/FileContentViewer.tsx:60]: add error handling if language is missing or null?

File: apps/gitness/src/components/FileViewer.tsx (hunk 2 above)
Comment (shaurya-harness) [reply apps/gitness/src/components/FileViewer.tsx:33]: sure

File: apps/gitness/src/utils/git-utils.ts (hunk 1 above)
Comment (shaurya-harness) [reply apps/gitness/src/utils/git-utils.ts:124]: sure

File: apps/gitness/src/components/FileContentViewer.tsx (hunk 3 above)
Comment (shaurya-harness) [reply apps/gitness/src/components/FileContentViewer.tsx]: the getDefaultView takes care of that

**Review by vardanbansal-harness — APPROVED** [review]

**Pull request merged**